### PR TITLE
Refactor to use `pkg/runtime` instead of direct calls to Docker

### PR DESCRIPTION
**Pull request by twelho**
This removes all direct calls to the `docker` binary in favor of the `pkg/runtime` interface.
- [x] `pkg` refactored
- [x] `cmd` refactored (excluding `attach`)
- [x] Global providers
- [x] Rebased on top of master

### CODE REVIEW

File: pkg/operations/start.go [hunk 1]
```diff
@@ -20,8 +21,14 @@ import (
 )
 
 func StartVM(vm *vmmd.VM, debug bool) error {
+	// Get the Docker client
+	dc, err := docker.GetDockerClient()
```
Comment (luxas) [inline pkg/operations/start.go]: make all these calls use the global provider

File: pkg/runtime/docker/client.go [hunk 2]
```diff
@@ -49,6 +56,102 @@ func (dc *dockerClient) PullImage(image string) (io.ReadCloser, error) {
...
+		AttachStdin:     true,
+		AttachStdout:    true,
+		AttachStderr:    true,
+		ExposedPorts:    nil,
+		Tty:             true,
+		OpenStdin:       true,
+		StdinOnce:       false,
+		Env:             nil,
+		Cmd:             nil,
+		Healthcheck:     nil,
```
Comment (luxas) [inline pkg/runtime/docker/client.go]: remove the unset options here to make sure the default values are used, and make it explicit what we set

File: pkg/source/docker.go [hunk 3]
```diff
@@ -66,34 +65,29 @@ func (ds *DockerSource) Parse(ociRef meta.OCIImageRef) (*api.OCIImageSource, err
...
-	if ds.containerID, err = util.ExecuteCommand("docker", "create", ds.imageID, "sh"); err != nil {
-		return nil, fmt.Errorf("failed to create Docker container from image %q: %v", ds.imageID, err)
-	}
-
-	// Open a tar file stream to be extracted straight into the image volume
-	ds.exportCmd = exec.Command("docker", "export", ds.containerID)
-	reader, err := ds.exportCmd.StdoutPipe()
+func (ds *DockerSource) Reader() (rc io.ReadCloser, err error) {
+	// Get the Docker client
+	dc, err := docker.GetDockerClient()
```
Comment (luxas) [inline pkg/source/docker.go]: global provider

File: pkg/operations/start.go (hunk 1 above)
Comment (twelho) [reply pkg/operations/start.go]: :+1:

File: pkg/runtime/docker/client.go (hunk 2 above)
Comment (twelho) [reply pkg/runtime/docker/client.go]: Yes, they were just temporarily there while I figured out which ones to set.

File: pkg/source/docker.go (hunk 3 above)
Comment (twelho) [reply pkg/source/docker.go]: :+1:

**Comment by twelho** [comment]
Docker client's  `ContainerAttach` and its `HijackedResponse` have quite lacking documentation, it seems like we need to implement a pseudo-TTY from scrach with signal handling, which is out of scope for this refactor. https://github.com/fsouza/go-dockerclient has an [implementation](https://github.com/fsouza/go-dockerclient/blob/cb91b9d898e126e9c5efb61491d90f39c0e03391/client.go#L716) for attach stream hijacking, but we'll stick to calling `docker attach` for now.

File: cmd/ignite/ignite.go [hunk 4]
```diff
@@ -4,6 +4,8 @@ import (
 	"fmt"
 	"os"
 
+	"github.com/weaveworks/ignite/pkg/providers"
+
```
Comment (luxas) [inline cmd/ignite/ignite.go]: remove blank line

File: cmd/ignite-spawn/ignite-spawn.go [hunk 5]
```diff
@@ -5,6 +5,8 @@ import (
 	"os"
 	"path"
 
+	"github.com/weaveworks/ignite/pkg/providers"
+
```
Comment (luxas) [inline cmd/ignite-spawn/ignite-spawn.go]: remove blank line

File: cmd/ignite/run/create.go [hunk 6]
```diff
@@ -5,10 +5,11 @@ import (
 	"path"
 	"strings"
 
+	"github.com/weaveworks/ignite/pkg/providers"
+
```
Comment (luxas) [inline cmd/ignite/run/create.go]: remove blank line

File: cmd/ignite/run/rmi.go [hunk 7]
```diff
@@ -4,7 +4,8 @@ import (
 	"fmt"
 	"os"
 
-	"github.com/weaveworks/ignite/pkg/client"
+	"github.com/weaveworks/ignite/pkg/providers"
+
```
Comment (luxas) [inline cmd/ignite/run/rmi.go]: remove blank line

File: cmd/ignite/run/rm.go [hunk 8]
```diff
@@ -3,7 +3,8 @@ package run
 import (
 	"fmt"
 
-	"github.com/weaveworks/ignite/pkg/client"
+	"github.com/weaveworks/ignite/pkg/providers"
+
```
Comment (luxas) [inline cmd/ignite/run/rm.go]: remove blank line

File: cmd/ignite/run/rmk.go [hunk 9]
```diff
@@ -4,7 +4,8 @@ import (
 	"fmt"
 	"os"
 
-	"github.com/weaveworks/ignite/pkg/client"
+	"github.com/weaveworks/ignite/pkg/providers"
+
```
Comment (luxas) [inline cmd/ignite/run/rmk.go]: remove blank line

File: pkg/metadata/id.go [hunk 10]
```diff
@@ -4,7 +4,8 @@ import (
 	"fmt"
 	"log"
 
-	"github.com/weaveworks/ignite/pkg/client"
+	"github.com/weaveworks/ignite/pkg/providers"
+
```
Comment (luxas) [inline pkg/metadata/id.go]: remove blank line

File: pkg/metadata/imgmd/imgmd.go [hunk 11]
```diff
@@ -3,6 +3,8 @@ package imgmd
 import (
 	"path"
 
+	"github.com/weaveworks/ignite/pkg/providers"
+
```
Comment (luxas) [inline pkg/metadata/imgmd/imgmd.go]: remove blank line

File: pkg/metadata/kernmd/kernmd.go [hunk 12]
```diff
@@ -3,6 +3,8 @@ package kernmd
 import (
 	"path"
 
+	"github.com/weaveworks/ignite/pkg/providers"
+
```
Comment (luxas) [inline pkg/metadata/kernmd/kernmd.go]: remove blank line

File: pkg/metadata/metadata.go [hunk 13]
```diff
@@ -7,6 +7,8 @@ import (
 	"path"
 	"regexp"
 
+	"github.com/weaveworks/ignite/pkg/providers"
+
```
Comment (luxas) [inline pkg/metadata/metadata.go]: remove blank line

File: pkg/metadata/vmmd/vmmd.go [hunk 14]
```diff
@@ -4,6 +4,8 @@ import (
 	"fmt"
 	"path"
 
+	"github.com/weaveworks/ignite/pkg/providers"
+
```
Comment (luxas) [inline pkg/metadata/vmmd/vmmd.go]: remove blank line

File: pkg/operations/start.go [hunk 15]
```diff
@@ -7,21 +7,22 @@ import (
 	"log"
 	"path"
 	"path/filepath"
-	"strings"
 	"time"
 
+	"github.com/weaveworks/ignite/pkg/providers"
+
```
Comment (luxas) [inline pkg/operations/start.go]: remove two blank lines

File: pkg/providers/network.go [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+
+import (
+	"github.com/weaveworks/ignite/pkg/network/cni"
+)
+
+// NetworkPlugin provides the default network plugin implementation
+var NetworkPlugin cni.NetworkPlugin
+
+func SetCNINetworkPlugin() (err error) {
+	NetworkPlugin, err = cni.GetCNINetworkPlugin(Runtime)
```
Comment (luxas) [inline pkg/providers/network.go]: this is ok, but a bit misleading, as by default we don't use this provider in the codepath at all

File: pkg/source/docker.go [hunk 17]
```diff
@@ -7,12 +7,13 @@ import (
 	"log"
 	"os/exec"
 
+	"github.com/weaveworks/ignite/pkg/providers"
+
```
Comment (luxas) [inline pkg/source/docker.go]: remove blank line

File: pkg/runtime/types.go [hunk 18]
```diff
@@ -1,16 +1,46 @@
...
+	AutoRemove   bool
+	NetworkMode  string
+	PortBindings meta.PortMappings
+}
+
 type Interface interface {
 	InspectImage(image string) (*ImageInspectResult, error)
 	PullImage(image string) (io.ReadCloser, error)
+	ExportImage(image string) (io.ReadCloser, string, error)
 	GetNetNS(containerID string) (string, error)
```
Comment (luxas) [inline pkg/runtime/types.go]: rename this to `GetContainerNetns()` and move to the container section?

File: pkg/runtime/types.go [hunk 19]
```diff
@@ -1,16 +1,46 @@
...
 	RawClient() interface{}
```
Comment (luxas) [inline pkg/runtime/types.go]: now that we do grouping here, move this last in its own section

File: pkg/runtime/types.go [hunk 20]
```diff
@@ -1,16 +1,46 @@
...
+
+	RunContainer(image string, config *ContainerConfig, name string) (string, error)
```
Comment (luxas) [inline pkg/runtime/types.go]: Add `// TODO: AttachContainer()`

File: pkg/operations/start.go (hunk 15 above)
Comment (twelho) [reply pkg/operations/start.go]: Will do a `make tidy` after the rebase that will get rid of the blank lines.

File: pkg/providers/network.go (hunk 16 above)
Comment (twelho) [reply pkg/providers/network.go]: It's currently only used on one place when using CNI networking, but at least it has its own provider now. Might rename it later if it causes extra confusion.

**Comment by twelho** [comment]
FYI: made this regex: `import \([^\)]*\n\n[^\)]*\n\n[^\)]*^\)` and fixed all excess blank lines in imports in the entire code in 6478c65.

File: pkg/source/docker.go (hunk 17 above)
Comment (twelho) [reply pkg/source/docker.go]: Done in 6478c65.

File: pkg/metadata/vmmd/vmmd.go (hunk 14 above)
Comment (twelho) [reply pkg/metadata/vmmd/vmmd.go]: Done in 6478c65.

File: pkg/metadata/metadata.go (hunk 13 above)
Comment (twelho) [reply pkg/metadata/metadata.go]: Done in 6478c65.

File: pkg/metadata/kernmd/kernmd.go (hunk 12 above)
Comment (twelho) [reply pkg/metadata/kernmd/kernmd.go]: Done in 6478c65.

File: pkg/metadata/imgmd/imgmd.go (hunk 11 above)
Comment (twelho) [reply pkg/metadata/imgmd/imgmd.go]: Done in 6478c65.

File: pkg/metadata/id.go (hunk 10 above)
Comment (twelho) [reply pkg/metadata/id.go]: Done in 6478c65.

File: cmd/ignite/run/rmk.go (hunk 9 above)
Comment (twelho) [reply cmd/ignite/run/rmk.go]: Done in 6478c65.

File: cmd/ignite/run/rm.go (hunk 8 above)
Comment (twelho) [reply cmd/ignite/run/rm.go]: Done in 6478c65.

File: cmd/ignite/run/rmi.go (hunk 7 above)
Comment (twelho) [reply cmd/ignite/run/rmi.go]: Done in 6478c65.

File: cmd/ignite/run/create.go (hunk 6 above)
Comment (twelho) [reply cmd/ignite/run/create.go]: Done in 6478c65.

File: cmd/ignite-spawn/ignite-spawn.go (hunk 5 above)
Comment (twelho) [reply cmd/ignite-spawn/ignite-spawn.go]: Done in 6478c65.

File: cmd/ignite/ignite.go (hunk 4 above)
Comment (twelho) [reply cmd/ignite/ignite.go]: Done in 6478c65.

File: pkg/operations/start.go (hunk 15 above)
Comment (twelho) [reply pkg/operations/start.go]: Done in 6478c65.

File: pkg/runtime/types.go (hunk 20 above)
Comment (twelho) [reply pkg/runtime/types.go]: Done in 6e60eba.

File: pkg/runtime/types.go (hunk 19 above)
Comment (twelho) [reply pkg/runtime/types.go]: Done in 6e60eba.

File: pkg/runtime/types.go (hunk 18 above)
Comment (twelho) [reply pkg/runtime/types.go]: Done in 6e60eba.

**Pull request merged**